src/adb.js

```diff
@@ -1 +1,353 @@
+			});
+	}
+
+	/**
+	 * Returns the pid of the specified app and device/emulator, if running.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} appid - The application's id
+	 * @returns {Promise}
+	 */
+	getPid(deviceId, appid) {
+		return this.shell(deviceId, 'ps')
+			.then(data => {
+				if (data) {
+					const lines = data.toString().split('\n');
+					for (let i = 0, j = lines.length; i < j; i++ ) {
+						let columns = lines[i].trim().split(/\s+/);
+						if (columns.pop() === appid) {
+							return parseInt(columns[1]);
+						}
+					}
+				}
+				return 0;
+			})
+			.catch(err => {
+				console.log('adb getPid: ', err);
+			});
+	}
+
+	/**
+	 * Installs an app to the specified device/emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} apkFile - The application apk file to install
+	 * @param {Object} [opts] - Install options
+	 * @param {Object} [opts.logger] - A logger instance
+	 * @returns {Promise}
+	 */
+	installApp(deviceId, apkFile, opts = {}) {
+		apkFile = path.resolve(apkFile);
+		if (!fs.existsSync(apkFile)) {
+			return Promise.reject('APK file does not exist', apkFile);
+		}
+
+		return this.devices()
+			.then(devices => {
+				if (devices.filter(d => { return d.id == deviceId; }).length != 1) {
+					return Promise.reject('device not found');
+				}
+
+				return sdk.detect();
+			})
+			.then(result => {
+				console.log(result);
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'install', '-r', apkFile]);
+			});
+			// .then(({ code, stdout, stderr }) => {})
+	}
+
+	/**
+	 * Starts an application on the specified device/emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} appid - The application's id
+	 * @param {String} activity - The name of the activity to run
+	 * @returns {Promise}
+	 */
+	startApp(deviceId, appid, activity) {
+		return this.shell(deviceId, 'am start -n ' + appid + '/.' + activity.replace(/^\./, ''));
+	}
+
+	/**
+	 * Stops an application on the specified device/emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} appid - The application's id
+	 * @returns {Promise}
+	 */
+	stopApp(deviceId, appid) {
+		return this.getPid(deviceId, appid)
+			.then(pid => {
+				console.log(pid);
+				return this.shell(deviceId, 'am force-stop ' + appid);
+			});
+	}
+
+	/**
+	 * Retrieves a list of connected devices and emulators.
+	 * @returns {Promise}
+	 */
+	devices() {
+		return new Connection(this)
+			.exec('host:devices')
+			.then(data => {
+				return parseDevices(this, data);
+			});
+	}
+
+	/**
+	 * Streams output from logcat into the specified handler until the adb logcat
+	 * process ends.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {Function} handler - A function to call whenever data becomes available
+	 * @returns {Promise}
+	 */
+	logcat(deviceId, handler) {
```

Comment:
Ensure that `deviceId` is a non-empty string and that handler is a function.

    typeof handler !== 'function'